Logs and metrics delivery via Stack Monitoring documentation (#4677)
elastic/cloud-on-k8s: docs/advanced-topics/stack-monitoring.asciidoc

:page_id: stack-monitoring
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]

[id="{p}-{page_id}"]
= Stack Monitoring

You can enable link:https://www.elastic.co/guide/en/elasticsearch/reference/current/monitor-elasticsearch-cluster.html[Stack Monitoring]
on your Elasticsearch and Kibana to collect and ship their metrics and logs.

> You just have to reference the separate monitoring Elasticsearch cluster in the `spec.monitoring` section.

Reviewer: ```suggestion
You just have to reference a separate monitoring Elasticsearch cluster in the `spec.monitoring` section.
```

Reviewer 2: Can a user configure the same Elasticsearch cluster to ship logs and metrics? This is a pattern we currently support in ECE, just curious, if the cluster has to be a separate cluster?

Author: A user cannot configure an Elasticsearch cluster to monitor itself, the monitoring cluster has to be a separate cluster.
This is not yet supported due to a circular dependency issue (https://github.com/elastic/cloud-on-k8s/issues/4627).

Author: I'm adding a note on this limitation.

[source,yaml,subs="attributes,callouts"]
----
cat $$<<$$EOF | kubectl apply -f -
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: monitored-sample
  namespace: production
spec:
  version: {version}
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  nodeSets:
  - name: default
    count: 1
    config:
      node.store.allow_mmap: false
---
apiVersion: kibana.k8s.elastic.co/{eck_crd_version}
kind: Kibana
metadata:
  name: monitored-sample
  namespace: production
spec:
  version: {version}
  elasticsearchRef:
    name: monitored-sample
    namespace: production
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  count: 1
EOF
----

> The use of `namespace` is optional if the Elasticsearch cluster and the Kibana are running in the same namespace.

Reviewer: Should this be a [callout ](https://github.com/elastic/docs#callouts) in the YAML sample?

Author: Good idea.

The monitoring cluster must be managed by ECK in the same Kubernetes cluster as the monitored one.

You can send metrics and logs to two different Elasticsearch monitoring clusters.

> You can enable Stack Monitoring only on Elasticsearch or only on Kibana. In the latter case, Kibana will not be available on the Stack Monitoring Kibana page (see link:https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data(View monitoring data in Kibana)).

Reviewer: ```suggestion
You can also enable Stack Monitoring only on Elasticsearch or only on Kibana. In the latter case, Kibana will not be available on the Stack Monitoring Kibana page (see link:https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data(View monitoring data in Kibana)).
```

== How it works

> In the background, Metricbeat and Filebeat are deployed as sidecar containers in the same Pod than Elasticsearch and Kibana.

Reviewer: ```suggestion
In the background, Metricbeat and Filebeat are deployed as sidecar containers in the same Pod as Elasticsearch and Kibana.
```

> Metricbeat is used to collect monitoring metrics and Filebeat to monitor the Elasticsearch log files, collect log events. The two beats are configured to ship data directly to the monitoring cluster(s) using dedicated Elastic users managed by ECK.

Reviewer: ```suggestion
Metricbeat is used to collect monitoring metrics and Filebeat to monitor the Elasticsearch log files, collect log events. The two Beats are configured to ship data directly to the monitoring cluster(s) using dedicated Elastic users managed by ECK.
```

> Elasticsearch and Kibana containers are configured to share the

Reviewer: Incomplete sentence?

Author: Whoooops. I was about to explain the logs volume sharing for Filebeat. But I'm not sure how far to go in explaining how it works. It seems to me that mentioning the sidecars is the most important part and the rest is secondary.

== Audit logging

Audit logs are collected and ship to the monitoring cluster referenced in the `monitoring.logs` section when audit logging is enabled (disabled by default).

[source,yaml,subs="attributes,callouts"]
----
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
spec:
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  nodeSets:
  - name: default
    config:
      # https://www.elastic.co/guide/en/elasticsearch/reference/current/enable-audit-logging.html
      xpack.security.audit.enabled: true
---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
spec:
  monitoring:
    metrics:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
    logs:
      elasticsearchRefs:
        - name: monitoring
          namespace: observability
  config:
    # https://www.elastic.co/guide/en/kibana/current/xpack-security-audit-logging.html
    xpack.security.audit.enabled: true
----

> == Override Beats Pod Template

Reviewer: ```suggestion
== Override the Beats Pod Template
```

> You can customize the Filebeat and Metricbeat containers through the Pod template. Your config is merged with our defaults.

Reviewer: ```suggestion
You can customize the Filebeat and Metricbeat containers through the Pod template. Your config is merged with  the values of the default Pod template ECK uses.
```

:page_id: stack-monitoring
ifdef::env-github[]
****
link:https://www.elastic.co/guide/en/cloud-on-k8s/master/k8s-{page_id}.html[View this document on the Elastic website]
****
endif::[]

[id="{p}-{page_id}"]
= Stack Monitoring

You can enable link:https://www.elastic.co/guide/en/elasticsearch/reference/current/monitor-elasticsearch-cluster.html[Stack Monitoring]

> on your Elasticsearch and Kibana to collect and ship their metrics and logs.

Reviewer 3: ```suggestion
on Elasticsearch and Kibana to collect and ship their metrics and logs to a dedicated monitoring cluster.
```
In the ESS docs we emphasize "the dedicated monitoring cluster" so maybe we could use that here too.

Author: Sounds good to me.

> You just have to reference a separate monitoring Elasticsearch cluster in the `spec.monitoring` section.

Reviewer 3: ```suggestion
To enable stack monitoring, simply reference the monitoring Elasticsearch cluster in the spec.monitoring section of the ECK configuration.
```

Author: `of the ECK configuration` is not really correct here. The ECK operator has a [configuration](https://www.elastic.co/guide/en/cloud-on-k8s/current/k8s-operator-config.html). Below this sentence, the YAML code corresponds to the Elasticsearch cluster specification and the Kibana instance(s) specification.

As we just said ES & KB in "You can enable Stack Monitoring on your Elasticsearch and Kibana to collect and ship their metrics and logs.", can we say:

To enable stack monitoring, simply reference the monitoring Elasticsearch cluster in the `spec.monitoring` section of their specification.

Reviewer 3: Thanks. I wasn't sure about that configuration part. Your suggested text looks perfect.

[source,yaml,subs="attributes,callouts"]
----
apiVersion: elasticsearch.k8s.elastic.co/{eck_crd_version}
kind: Elasticsearch
metadata:
  name: monitored-sample
  namespace: production
spec:
  version: {version}
  monitoring:
    metrics:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability <1>
    logs:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability <1>
  nodeSets:
  - name: default
    count: 1
    config:
      node.store.allow_mmap: false
---
apiVersion: kibana.k8s.elastic.co/{eck_crd_version}
kind: Kibana
metadata:
  name: monitored-sample
  namespace: production
spec:
  version: {version}
  elasticsearchRef:
    name: monitored-sample
    namespace: production <2>
  monitoring:
    metrics:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability <1>
    logs:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability <1>
  count: 1
----

<1> The use of `namespace` is optional if the monitoring Elasticsearch cluster and the monitored Elasticsearch cluster are running in the same namespace.

> <2> The use of `namespace` is optional if the Elasticsearch cluster and the Kibana are running in the same namespace.

Reviewer 3: ```suggestion
<2> The use of `namespace` is optional if the Elasticsearch cluster and the Kibana instance are running in the same namespace.
```

CAUTION: You cannot configure an Elasticsearch cluster to monitor itself, the monitoring cluster has to be a separate cluster.

IMPORTANT: The monitoring cluster must be managed by ECK in the same Kubernetes cluster as the monitored one.

You can send metrics and logs to two different Elasticsearch monitoring clusters.

> You can also enable Stack Monitoring only on Elasticsearch or only on Kibana. In the latter case, Kibana will not be available on the Stack Monitoring Kibana page (see link:https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data[View monitoring data in Kibana]).

Reviewer 3: ```suggestion
You can also enable Stack Monitoring on Elasticsearch only or on Kibana only. In the latter case, Kibana will not be available on the Stack Monitoring Kibana page (see link:https://www.elastic.co/guide/en/kibana/current/monitoring-data.html#monitoring-data[View monitoring data in Kibana]).
```

== When to use it

This feature is a good solution if you need to monitor your Elastic applications in restricted Kubernetes environments where you cannot grant advanced permissions:

- to Metricbeat to allow queriying the k8s API
- to Filebeat to deploy a privileged DaemonSet

However, for maximum efficiency and minimising resource consumption, or advanced use cases that require specific Beats configurations, you can deploy a standalone Metricbeat Deployment and a Filebeat Daemonset. See the <<{p}-beat-configuration-examples,Beats configuration Examples>> for more information.

== How it works

In the background, Metricbeat and Filebeat are deployed as sidecar containers in the same Pod as Elasticsearch and Kibana.

Metricbeat is used to collect monitoring metrics and Filebeat to monitor the Elasticsearch log files and collect log events.

The two Beats are configured to ship data directly to the monitoring cluster(s) using HTTPS and dedicated Elastic users managed by ECK.

== Audit logging

> Audit logs are collected and ship to the monitoring cluster referenced in the `monitoring.logs` section when audit logging is enabled (disabled by default).

Reviewer 3: ```suggestion
Audit logs are collected and shipped to the monitoring cluster referenced in the `monitoring.logs` section when audit logging is enabled (disabled by default).
```

[source,yaml,subs="attributes,callouts"]
----
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
spec:
  monitoring:
    metrics:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability
    logs:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability
  nodeSets:
  - name: default
    config:
      # https://www.elastic.co/guide/en/elasticsearch/reference/current/enable-audit-logging.html
      xpack.security.audit.enabled: true
---
apiVersion: kibana.k8s.elastic.co/v1
kind: Kibana
spec:
  monitoring:
    metrics:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability
    logs:
      elasticsearchRefs:
      - name: monitoring
        namespace: observability
  config:
    # https://www.elastic.co/guide/en/kibana/current/xpack-security-audit-logging.html
    xpack.security.audit.enabled: true
----

== Override the Beats Pod Template

> You can customize the Filebeat and Metricbeat containers through the Pod template. Your config is merged with the values of the default Pod template ECK uses.

Reviewer 3: ```suggestion
You can customize the Filebeat and Metricbeat containers through the Pod template. Your configuration is merged with the values of the default Pod template that ECK uses.
```
